src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -631,18 +632,24 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
                 Debug.Assert(((VariableDeclarationSyntax)topLevelVariableDeclaration).Deconstruction.Value != null);
 
                 var statement = topLevelVariableDeclaration.Parent;
+                TypeSymbol result;
                 switch (statement.Kind())
                 {
                     case SyntaxKind.LocalDeclarationStatement:
                         this.binder.BindLocalDeclarationStatement((LocalDeclarationStatementSyntax)statement, diagnostics);
+                        break;
 
-                        TypeSymbol result = this._type;
-                        Debug.Assert((object)result != null);
-                        return result;
+                    case SyntaxKind.ForStatement:
+                        this.binder.BindFor((ForStatementSyntax)statement, diagnostics);
```

Comment:
You're right, we should be able to bind just the declaration expression here. 
The trouble is that needs to be done in the right binder. I'll look into this.